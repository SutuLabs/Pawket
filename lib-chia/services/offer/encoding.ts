import { CoinSpend, OriginCoin, SpendBundle } from "../spendbundle";
import { bech32m } from "@scure/base";
import zlib from "zlib";
import { Buffer } from "buffer";
import { Bytes, sexp_buffer_from_stream, Stream, sexp_from_stream, SExp } from "clvm";
import { prefix0x } from "../coin/condition";
import { assemble, disassemble } from "clvm_tools/clvm_tools/binutils";
import puzzle from "../crypto/puzzle";
import { modshex } from "../coin/mods";

// sync with `chia\wallet\util\puzzle_compression.py`
const initDict = [
  modshex["p2_delegated_puzzle_or_hidden_puzzle"] + modshex["cat_v1"],
  modshex["settlement_payments"],
  modshex["singleton_top_layer_v1_1"] +
    modshex["nft_state_layer"] +
    modshex["nft_ownership_layer"] +
    modshex["nft_metadata_updater_default"] +
    modshex["nft_ownership_transfer_program_one_way_claim_with_royalties"],
  modshex["cat_v2"],
  modshex["settlement_payments_v1"],
  "", // chia client purposefully break version, to prevent legacy client accept new offer: Immediate error rather than on chain failure.
].map((t) => Buffer.from(t, "hex"));

function getDictForVersion(ver: number) {
  let b = Buffer.from([]);
  for (const r of initDict.slice(0, ver)) b = Buffer.concat([b, r]);
  return b;
}

export async function decodeOffer(offerText: string): Promise<SpendBundle> {
  const offer_compressed = bech32m.decodeToBytes(offerText).bytes;
  const buff = Buffer.from(offer_compressed);

  const ver = buff.readUInt16BE(0);
  if (ver > initDict.length) throw new Error("error offer file version");

  const d = zlib.inflateSync(buff.slice(2), { dictionary: getDictForVersion(ver) });

  let pos = 0;
  const dv = new DataView(d.buffer);
  const spendnum = dv.getUint32(pos, false);
  pos += 32 / 8;
  const spends = [];
  for (let i = 0; i < spendnum; i++) {
    const pci = Bytes.from(d.slice(pos, pos + 32)).hex();
    pos += 32;
    const ph = Bytes.from(d.slice(pos, pos + 32)).hex();
    pos += 32;
    const amount = dv.getBigUint64(pos, false);
    pos += 64 / 8;
    const coin: OriginCoin = {
      parent_coin_info: prefix0x(pci),
      puzzle_hash: prefix0x(ph),
      amount: amount,
    };

    const readSexpBuffer = () => {
      const buff = d.slice(pos);
      const r = sexp_buffer_from_stream(new Stream(Bytes.from(new Uint8Array(buff))));
      const n = r.raw().length;
      pos += n;
      return disassemble(sexp_from_stream(new Stream(r), SExp.to));
    };

    const puzzle_reveal = readSexpBuffer();
    const solution = readSexpBuffer();

    const coinspend: CoinSpend = {
      coin: coin,
      puzzle_reveal: prefix0x(await puzzle.encodePuzzle(puzzle_reveal)),
      solution: prefix0x(await puzzle.encodePuzzle(solution)),
    };

    spends.push(coinspend);
  }

  const sig = Bytes.from(d.slice(pos, pos + 96)).hex();
  pos += 96;

  const bundle: SpendBundle = {
    aggregated_signature: prefix0x(sig),
    coin_spends: spends,
  };

  return bundle;
}

export async function encodeOffer(bundle: SpendBundle, ver: number | undefined = undefined, prefix = "offer"): Promise<string> {
  // big endian
  const chunks: Buffer[] = [];
  const spendnum = bundle.coin_spends.length;
  chunks.push(getUint32Buffer(spendnum));

  for (let i = 0; i < spendnum; i++) {
    const cs = bundle.coin_spends[i];
    const coin = cs.coin;
    chunks.push(Buffer.from(Bytes.from(coin.parent_coin_info, "hex").raw()));
    chunks.push(Buffer.from(Bytes.from(coin.puzzle_hash, "hex").raw()));
    chunks.push(getUint64Buffer(coin.amount));

    chunks.push(
      Buffer.from(
        assemble(await puzzle.disassemblePuzzle(cs.puzzle_reveal))
          .as_bin()
          .raw()
      )
    );
    chunks.push(
      Buffer.from(
        assemble(await puzzle.disassemblePuzzle(cs.solution))
          .as_bin()
          .raw()
      )
    );
  }

  chunks.push(Buffer.from(Bytes.from(bundle.aggregated_signature, "hex").raw()));

  const buff = Buffer.concat(chunks);

  ver = ver ?? initDict.length;

  const def = zlib.deflateSync(buff, { dictionary: getDictForVersion(ver) });

  const final_buff = Buffer.concat([getUint16Buffer(ver), def]);

  const encoded = bech32m.encode(prefix, bech32m.toWords(final_buff), false);
  return encoded;
}

function getUint16Buffer(number: number): Buffer {
  const buf = Buffer.alloc(2);
  buf.writeUInt16BE(number);
  return buf;
}

function getUint32Buffer(number: number): Buffer {
  const buf = Buffer.alloc(4);
  buf.writeUInt32BE(number);
  return buf;
}

function getUint64Buffer(number: bigint): Buffer {
  const buf = Buffer.alloc(8);
  writeBigUInt64BE(buf, number);
  return buf;
}

// due to unknown `writeBigUInt64BE not a function` error, get following code from https://github.com/feross/buffer/blob/master/index.js#L1516
function writeBigUInt64BE(buf: Buffer, value: bigint, offset = 0) {
  return wrtBigUInt64BE(buf, value, offset);
}

function wrtBigUInt64BE(buf: Buffer, value: bigint, offset: number) {
  let lo = Number(value & BigInt(0xffffffff));
  buf[offset + 7] = lo;
  lo = lo >> 8;
  buf[offset + 6] = lo;
  lo = lo >> 8;
  buf[offset + 5] = lo;
  lo = lo >> 8;
  buf[offset + 4] = lo;
  let hi = Number((value >> BigInt(32)) & BigInt(0xffffffff));
  buf[offset + 3] = hi;
  hi = hi >> 8;
  buf[offset + 2] = hi;
  hi = hi >> 8;
  buf[offset + 1] = hi;
  hi = hi >> 8;
  buf[offset] = hi;
  return offset + 8;
}
