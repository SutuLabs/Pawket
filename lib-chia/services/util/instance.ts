import { ModuleInstance } from "@chiamine/bls-signatures";
import loadBls from "@chiamine/bls-signatures";
import * as clvm_tools from "clvm_tools";

class InstanceSingleton {
  private static _instance: InstanceSingleton = new InstanceSingleton();

  BLS?: ModuleInstance;

  constructor() {
    if (InstanceSingleton._instance) {
      throw new Error("Error: Instantiation failed: Use InstanceSingleton.getInstance() instead of new.");
    }
    InstanceSingleton._instance = this;
  }

  public static getInstance(): InstanceSingleton {
    return InstanceSingleton._instance;
  }

  public async init(): Promise<void> {
    this.BLS = await loadBls();
    await clvm_tools.initialize();
  }
}

export const Instance = InstanceSingleton.getInstance();
