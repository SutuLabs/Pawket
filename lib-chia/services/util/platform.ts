export function isIos(): boolean {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  if ((<any>window).cordova && (<any>window).cordova.platformId == "ios") return true;
  return false;
}
