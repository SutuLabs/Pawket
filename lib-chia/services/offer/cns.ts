import { UnsignedSpendBundle, combineSpendBundle, OriginCoin } from "../spendbundle";
import { SymbolCoins } from "../transfer/transfer";
import { analyzeNftCoin, generateMintNftBundle } from "../coin/nft";
import puzzle from "../crypto/puzzle";
import { CnsMetadataValues } from "../../models/nft";
import { getCoinName0x, NetworkContext } from "../coin/coinUtility";
import receive, { TokenPuzzleDetail } from "../crypto/receive";
import { generateOfferPlan, generateNftOffer } from "./bundler";
import { prefix0x } from "../coin/condition";
import { GetParentPuzzleResponse } from "../../models/api";
import { OfferEntity } from "./summary";
import utility from "../crypto/utility";
import { Instance } from "../util/instance";

export async function generateMintCnsOffer(
  targetAddress: string,
  changeAddress: string,
  price: bigint,
  fee: bigint,
  metadata: CnsMetadataValues,
  availcoins: SymbolCoins,
  requests: TokenPuzzleDetail[],
  royaltyAddressHex: string,
  tradePricePercentage: number,
  net: NetworkContext,
  nonceHex: string | null = null,
  privateKey: string | undefined = undefined
): Promise<UnsignedSpendBundle> {
  if (!metadata.expiry) throw new Error("Expiry date is mandatory for CNS.");
  if (!metadata.name) throw new Error("Name is mandatory for CNS.");

  const target_hex = prefix0x(puzzle.getPuzzleHashFromAddress(targetAddress));
  const change_hex = prefix0x(puzzle.getPuzzleHashFromAddress(changeAddress));
  const reqs = [
    {
      id: "",
      symbol: net.symbol,
      amount: price,
      target: target_hex,
    },
  ];

  // generate temporary intermediate address
  const BLS = Instance.BLS;
  if (!BLS) throw new Error("BLS is not initialized.");
  const sk = privateKey ? privateKey : utility.toHexString(BLS.AugSchemeMPL.key_gen(utility.getRandom(64)).serialize());
  const puzzles = await receive.getAssetsRequestDetail(sk, 0, 1, [], {}, "any", net.symbol, "cat_v2");
  const ps = puzzles.filter((_) => _.symbol == net.symbol)[0].puzzles;
  const intermediateAddress = ps[0].address;
  requests.filter((_) => _.symbol == net.symbol)[0].puzzles.push(...ps);

  const spendBundle = await generateMintNftBundle(
    intermediateAddress,
    changeAddress,
    fee,
    metadata,
    availcoins,
    requests,
    royaltyAddressHex,
    tradePricePercentage,
    net,
    undefined,
    undefined,
    undefined,
    true
  );

  const nftcs = spendBundle.coin_spends[2];
  const analysis = await analyzeNftCoin(nftcs.puzzle_reveal, "", nftcs.coin, nftcs.solution);
  const nextCoin: OriginCoin = {
    parent_coin_info: getCoinName0x(nftcs.coin),
    amount: 1n,
    puzzle_hash: nftcs.coin.puzzle_hash,
  };
  const knownCoins: GetParentPuzzleResponse[] = [];
  knownCoins.push({
    parentCoinId: nextCoin.parent_coin_info,
    parentParentCoinId: nftcs.coin.parent_coin_info,
    amount: Number(nftcs.coin.amount),
    puzzleReveal: nftcs.puzzle_reveal,
  });
  if (!analysis) throw Error("failed to analysis nft");
  if ("cnsName" in analysis === false) throw Error("failed to analysis cns nft");

  const offs: OfferEntity[] = [
    {
      id: analysis.launcherId,
      amount: 0n,
      royalty: analysis.tradePricePercentage,
      nft_uri: "something unimportant",
    },
  ];

  const offplan = await generateOfferPlan(offs, change_hex, availcoins, 0n, net.symbol);
  const legacyApiCall = net.api;
  net.api = async function (parentCoinId: string): Promise<GetParentPuzzleResponse | undefined> {
    const resp = knownCoins.find((_) => _.parentCoinId == parentCoinId);
    if (resp) return resp;
    return await legacyApiCall(parentCoinId);
  };
  const offerBundle = await generateNftOffer(offplan, analysis, nextCoin, reqs, requests, net, nonceHex);

  const offerCombineBundle = combineSpendBundle(spendBundle, offerBundle);
  return offerCombineBundle;
}
