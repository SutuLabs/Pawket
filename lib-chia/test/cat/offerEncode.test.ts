import { decodeOffer, encodeOffer } from "../../services/offer/encoding";
import { getOfferSummary } from "../../services/offer/summary";
import { Instance } from "../../services/util/instance";

beforeAll(async () => {
  await Instance.init();
});

test("Offer Encoding", async () => {
  const offerText =
    "offer1qqp83w76wzru6cmqvpsxygqqwc7hynr6hum6e0mnf72sn7uvvkpt68eyumkhelprk0adeg42nlelk2mpafrgx923m0l47c8udnalkndly6tx3nlj8kw87vhh5p4vekl6ul7ll0myae24ja0w6azxqnj67rchq77vsy0kw0w4d6k8fvjj7yc4s86alw454adyl0gcyeqhu4fc4c0lt0znmjj7xzq26y0m8mm7n6lesmnlfrj95u0nldhzag2m7yl0fhh7d8f708kwg6hn4h8nlh4clgclva2xuthwndt9n7e8hduamg36278zdl70wdmauhxg0sk7d8zgxc6gy3ks2tyjgcdj82pfvz6rzvk6rqvkmryvxerwqffpm5dz8crzdaqar0grf8knu4ew4ldrz6wpsj6gap0trj63jwt8azy57h6ckrmydaua8dgx2tst4amn0ud8krwtxwgjd5e9nv0zu7d8akufy2matea866ktqhc7hqw44zpzzlev8w2np5m83s754wfc5aa9z645j20lamxwc7smc9acuw7hdtrmwcx7lvazwt722jh4llph6rj5eg0xkugtarwt4cm3hz2rfv8mpggdjznvnlc9s3drnlumgx6jwct4ypp7mle0vp7tymdeun8uexqlpv0me9lm7xgfk0tk4adv3mjft3k8rvta9grwuyep38h0lqj456rjdmwsn05z6mlkdm6hrhkhqaktd0ranetjmvnwe7edhe7la5elaltlg8l8lcgfpj6uwlhxn6254hnn7zhj6cc5v0a88wgm9vt9k85axakerqkplutzddwt8vd0utcs33sl7fs3740z0cj4w8u5rf5m62da4u6av4fr399wkv7ylwlwaf4qhdyd2q479lczgtkssupsxsgxt9xz0tq4ckgrl8my5zw87v3gn36tt6780c05n0m7tukx5dtemj35xelcgxs9p2ypn3mexa8mpxkp6g2fxu2njl46fcyv4jcvdj958vma6ce0suj8p0wdafglydnv9edahwhe4kr5lcrza9etujm6g5t8m4jdtfffrcn080e363wdpfdm8a0ay7vrtl9ftl2pjw4p8lfmdxelwctc8tspul7u7zs8rqqqcc2wacptnhjp";
  const bundle = await decodeOffer(offerText);
  const summary = await getOfferSummary(bundle);

  expect(summary).toMatchSnapshot("summary");
  expect(bundle).toMatchSnapshot("bundle");

  const encoded = await encodeOffer(bundle, 2);
  expect(encoded).toBe(offerText);
});
