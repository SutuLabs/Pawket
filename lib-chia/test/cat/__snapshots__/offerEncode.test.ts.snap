// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offer Encoding: bundle 1`] = `
Object {
  "aggregated_signature": "0x98c70db3903a9804825c68d92d0ac911986a56815966dcd9475dd354600aed6a8c224b318376ad85b3ed484c51a31b770ed22e72388e3b822a1a5917ed896f41153671856b2f4f903680fc7fa28955908f693d8b0f5dba002d3009dff320c000",
  "coin_spends": Array [
    Object {
      "coin": Object {
        "amount": 0n,
        "parent_coin_info": "0x0000000000000000000000000000000000000000000000000000000000000000",
        "puzzle_hash": "0xbae24162efbd568f89bc7a340798a6118df0189eb9e3f8697bcea27af99f8f79",
      },
      "puzzle_reveal": "0xff02ffff01ff02ff0affff04ff02ffff04ff03ff80808080ffff04ffff01ffff333effff02ffff03ff05ffff01ff04ffff04ff0cffff04ffff02ff1effff04ff02ffff04ff09ff80808080ff808080ffff02ff16ffff04ff02ffff04ff19ffff04ffff02ff0affff04ff02ffff04ff0dff80808080ff808080808080ff8080ff0180ffff02ffff03ff05ffff01ff04ffff04ff08ff0980ffff02ff16ffff04ff02ffff04ff0dffff04ff0bff808080808080ffff010b80ff0180ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff1effff04ff02ffff04ff09ff80808080ffff02ff1effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080",
      "solution": "0xffffa033e6b7efb2fa91742b9a4dc36fc96dc1255ced7fe78ffff469bd24aacefa913bffffa0f1c3875cc0f8cd4926d8a9423a244c30af5b7dfa7a25c947c570634423ca2157ff85174876e800ff80808080",
    },
    Object {
      "coin": Object {
        "amount": 1337000n,
        "parent_coin_info": "0x16f79de5eb9f31ce2ec65b67c7c6a8d5d44ec8ef35dd6ecbc9c6c9ab37b5e7f1",
        "puzzle_hash": "0xd8d5f15deb6861bb977d76f26b8db6cabc1a21c8fd5ff7ec8e7703c256ee285f",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ff5effff04ff02ffff04ffff04ff05ffff04ffff0bff2cff0580ffff04ff0bff80808080ffff04ffff02ff17ff2f80ffff04ff5fffff04ffff02ff2effff04ff02ffff04ff17ff80808080ffff04ffff0bff82027fff82057fff820b7f80ffff04ff81bfffff04ff82017fffff04ff8202ffffff04ff8205ffffff04ff820bffff80808080808080808080808080ffff04ffff01ffffffff81ca3dff46ff0233ffff3c04ff01ff0181cbffffff02ff02ffff03ff05ffff01ff02ff32ffff04ff02ffff04ff0dffff04ffff0bff22ffff0bff2cff3480ffff0bff22ffff0bff22ffff0bff2cff5c80ff0980ffff0bff22ff0bffff0bff2cff8080808080ff8080808080ffff010b80ff0180ffff02ffff03ff0bffff01ff02ffff03ffff09ffff02ff2effff04ff02ffff04ff13ff80808080ff820b9f80ffff01ff02ff26ffff04ff02ffff04ffff02ff13ffff04ff5fffff04ff17ffff04ff2fffff04ff81bfffff04ff82017fffff04ff1bff8080808080808080ffff04ff82017fff8080808080ffff01ff088080ff0180ffff01ff02ffff03ff17ffff01ff02ffff03ffff20ff81bf80ffff0182017fffff01ff088080ff0180ffff01ff088080ff018080ff0180ffff04ffff04ff05ff2780ffff04ffff10ff0bff5780ff778080ff02ffff03ff05ffff01ff02ffff03ffff09ffff02ffff03ffff09ff11ff7880ffff0159ff8080ff0180ffff01818f80ffff01ff02ff7affff04ff02ffff04ff0dffff04ff0bffff04ffff04ff81b9ff82017980ff808080808080ffff01ff02ff5affff04ff02ffff04ffff02ffff03ffff09ff11ff7880ffff01ff04ff78ffff04ffff02ff36ffff04ff02ffff04ff13ffff04ff29ffff04ffff0bff2cff5b80ffff04ff2bff80808080808080ff398080ffff01ff02ffff03ffff09ff11ff2480ffff01ff04ff24ffff04ffff0bff20ff2980ff398080ffff010980ff018080ff0180ffff04ffff02ffff03ffff09ff11ff7880ffff0159ff8080ff0180ffff04ffff02ff7affff04ff02ffff04ff0dffff04ff0bffff04ff17ff808080808080ff80808080808080ff0180ffff01ff04ff80ffff04ff80ff17808080ff0180ffffff02ffff03ff05ffff01ff04ff09ffff02ff26ffff04ff02ffff04ff0dffff04ff0bff808080808080ffff010b80ff0180ff0bff22ffff0bff2cff5880ffff0bff22ffff0bff22ffff0bff2cff5c80ff0580ffff0bff22ffff02ff32ffff04ff02ffff04ff07ffff04ffff0bff2cff2c80ff8080808080ffff0bff2cff8080808080ffff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff2effff04ff02ffff04ff09ff80808080ffff02ff2effff04ff02ffff04ff0dff8080808080ffff01ff0bff2cff058080ff0180ffff04ffff04ff28ffff04ff5fff808080ffff02ff7effff04ff02ffff04ffff04ffff04ff2fff0580ffff04ff5fff82017f8080ffff04ffff02ff7affff04ff02ffff04ff0bffff04ff05ffff01ff808080808080ffff04ff17ffff04ff81bfffff04ff82017fffff04ffff0bff8204ffffff02ff36ffff04ff02ffff04ff09ffff04ff820affffff04ffff0bff2cff2d80ffff04ff15ff80808080808080ff8216ff80ffff04ff8205ffffff04ff820bffff808080808080808080808080ff02ff2affff04ff02ffff04ff5fffff04ff3bffff04ffff02ffff03ff17ffff01ff09ff2dffff0bff27ffff02ff36ffff04ff02ffff04ff29ffff04ff57ffff04ffff0bff2cff81b980ffff04ff59ff80808080808080ff81b78080ff8080ff0180ffff04ff17ffff04ff05ffff04ff8202ffffff04ffff04ffff04ff24ffff04ffff0bff7cff2fff82017f80ff808080ffff04ffff04ff30ffff04ffff0bff81bfffff0bff7cff15ffff10ff82017fffff11ff8202dfff2b80ff8202ff808080ff808080ff138080ff80808080808080808080ff018080ffff04ffff01a072dec062874cd4d3aab892a0906688a1ae412b0109982e1797a170add88bdcdcffff04ffff01a044a1d78b820f6404de3cc45b34932178f9ac8f3d9114db279f657ca83fa751b7ffff04ffff01ff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b095d59273ba6573261a23fb89698aed580dce0b6d3ab8f6530ddf2cc9d11c7bafa3fd5144cc7129f6540eb1a6d7833217ff018080ff0180808080",
      "solution": "0xffff80ffff01ffff33ffa0bae24162efbd568f89bc7a340798a6118df0189eb9e3f8697bcea27af99f8f79ff831466a8ffffa0bae24162efbd568f89bc7a340798a6118df0189eb9e3f8697bcea27af99f8f798080ffff3fffa001e4b6a719f9695cf8a1e0ef1ef8e10b075dbd2654bb920a065e5d8e243047918080ff8080ffffa0aa35c292b82e914b1f36eeabc70debc03cb4ede76e7b14aebca6dbcfeebe33fbffa06fff1390007dc6f96dcd1a63bcc1fd3d015a0167e71ed8a8a339c59e96a22030ff840226d3b880ffa0115c07f6185aef2c58ea7377c102d91bb2876f451c640822d704e766d745453dffffa016f79de5eb9f31ce2ec65b67c7c6a8d5d44ec8ef35dd6ecbc9c6c9ab37b5e7f1ffa0d8d5f15deb6861bb977d76f26b8db6cabc1a21c8fd5ff7ec8e7703c256ee285fff831466a880ffffa016f79de5eb9f31ce2ec65b67c7c6a8d5d44ec8ef35dd6ecbc9c6c9ab37b5e7f1ffa020af116d1d301f87791841cf1461c774e8ef44094e9ef4d3b1c5d5427628b3f1ff831466a880ff80ff8080",
    },
  ],
}
`;

exports[`Offer Encoding: summary 1`] = `
Object {
  "offered": Array [
    Object {
      "amount": 1337000n,
      "cat_target": "0xad61590f43f3621d21211d0f1d915449467d8e738b97c4f98de7fcb9a12b3ed5",
      "coin": Object {
        "coin": Object {
          "amount": 1337000n,
          "parent_coin_info": "0x16f79de5eb9f31ce2ec65b67c7c6a8d5d44ec8ef35dd6ecbc9c6c9ab37b5e7f1",
          "puzzle_hash": "0xd8d5f15deb6861bb977d76f26b8db6cabc1a21c8fd5ff7ec8e7703c256ee285f",
        },
        "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ff5effff04ff02ffff04ffff04ff05ffff04ffff0bff2cff0580ffff04ff0bff80808080ffff04ffff02ff17ff2f80ffff04ff5fffff04ffff02ff2effff04ff02ffff04ff17ff80808080ffff04ffff0bff82027fff82057fff820b7f80ffff04ff81bfffff04ff82017fffff04ff8202ffffff04ff8205ffffff04ff820bffff80808080808080808080808080ffff04ffff01ffffffff81ca3dff46ff0233ffff3c04ff01ff0181cbffffff02ff02ffff03ff05ffff01ff02ff32ffff04ff02ffff04ff0dffff04ffff0bff22ffff0bff2cff3480ffff0bff22ffff0bff22ffff0bff2cff5c80ff0980ffff0bff22ff0bffff0bff2cff8080808080ff8080808080ffff010b80ff0180ffff02ffff03ff0bffff01ff02ffff03ffff09ffff02ff2effff04ff02ffff04ff13ff80808080ff820b9f80ffff01ff02ff26ffff04ff02ffff04ffff02ff13ffff04ff5fffff04ff17ffff04ff2fffff04ff81bfffff04ff82017fffff04ff1bff8080808080808080ffff04ff82017fff8080808080ffff01ff088080ff0180ffff01ff02ffff03ff17ffff01ff02ffff03ffff20ff81bf80ffff0182017fffff01ff088080ff0180ffff01ff088080ff018080ff0180ffff04ffff04ff05ff2780ffff04ffff10ff0bff5780ff778080ff02ffff03ff05ffff01ff02ffff03ffff09ffff02ffff03ffff09ff11ff7880ffff0159ff8080ff0180ffff01818f80ffff01ff02ff7affff04ff02ffff04ff0dffff04ff0bffff04ffff04ff81b9ff82017980ff808080808080ffff01ff02ff5affff04ff02ffff04ffff02ffff03ffff09ff11ff7880ffff01ff04ff78ffff04ffff02ff36ffff04ff02ffff04ff13ffff04ff29ffff04ffff0bff2cff5b80ffff04ff2bff80808080808080ff398080ffff01ff02ffff03ffff09ff11ff2480ffff01ff04ff24ffff04ffff0bff20ff2980ff398080ffff010980ff018080ff0180ffff04ffff02ffff03ffff09ff11ff7880ffff0159ff8080ff0180ffff04ffff02ff7affff04ff02ffff04ff0dffff04ff0bffff04ff17ff808080808080ff80808080808080ff0180ffff01ff04ff80ffff04ff80ff17808080ff0180ffffff02ffff03ff05ffff01ff04ff09ffff02ff26ffff04ff02ffff04ff0dffff04ff0bff808080808080ffff010b80ff0180ff0bff22ffff0bff2cff5880ffff0bff22ffff0bff22ffff0bff2cff5c80ff0580ffff0bff22ffff02ff32ffff04ff02ffff04ff07ffff04ffff0bff2cff2c80ff8080808080ffff0bff2cff8080808080ffff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff2effff04ff02ffff04ff09ff80808080ffff02ff2effff04ff02ffff04ff0dff8080808080ffff01ff0bff2cff058080ff0180ffff04ffff04ff28ffff04ff5fff808080ffff02ff7effff04ff02ffff04ffff04ffff04ff2fff0580ffff04ff5fff82017f8080ffff04ffff02ff7affff04ff02ffff04ff0bffff04ff05ffff01ff808080808080ffff04ff17ffff04ff81bfffff04ff82017fffff04ffff0bff8204ffffff02ff36ffff04ff02ffff04ff09ffff04ff820affffff04ffff0bff2cff2d80ffff04ff15ff80808080808080ff8216ff80ffff04ff8205ffffff04ff820bffff808080808080808080808080ff02ff2affff04ff02ffff04ff5fffff04ff3bffff04ffff02ffff03ff17ffff01ff09ff2dffff0bff27ffff02ff36ffff04ff02ffff04ff29ffff04ff57ffff04ffff0bff2cff81b980ffff04ff59ff80808080808080ff81b78080ff8080ff0180ffff04ff17ffff04ff05ffff04ff8202ffffff04ffff04ffff04ff24ffff04ffff0bff7cff2fff82017f80ff808080ffff04ffff04ff30ffff04ffff0bff81bfffff0bff7cff15ffff10ff82017fffff11ff8202dfff2b80ff8202ff808080ff808080ff138080ff80808080808080808080ff018080ffff04ffff01a072dec062874cd4d3aab892a0906688a1ae412b0109982e1797a170add88bdcdcffff04ffff01a044a1d78b820f6404de3cc45b34932178f9ac8f3d9114db279f657ca83fa751b7ffff04ffff01ff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b095d59273ba6573261a23fb89698aed580dce0b6d3ab8f6530ddf2cc9d11c7bafa3fd5144cc7129f6540eb1a6d7833217ff018080ff0180808080",
        "solution": "0xffff80ffff01ffff33ffa0bae24162efbd568f89bc7a340798a6118df0189eb9e3f8697bcea27af99f8f79ff831466a8ffffa0bae24162efbd568f89bc7a340798a6118df0189eb9e3f8697bcea27af99f8f798080ffff3fffa001e4b6a719f9695cf8a1e0ef1ef8e10b075dbd2654bb920a065e5d8e243047918080ff8080ffffa0aa35c292b82e914b1f36eeabc70debc03cb4ede76e7b14aebca6dbcfeebe33fbffa06fff1390007dc6f96dcd1a63bcc1fd3d015a0167e71ed8a8a339c59e96a22030ff840226d3b880ffa0115c07f6185aef2c58ea7377c102d91bb2876f451c640822d704e766d745453dffffa016f79de5eb9f31ce2ec65b67c7c6a8d5d44ec8ef35dd6ecbc9c6c9ab37b5e7f1ffa0d8d5f15deb6861bb977d76f26b8db6cabc1a21c8fd5ff7ec8e7703c256ee285fff831466a880ffffa016f79de5eb9f31ce2ec65b67c7c6a8d5d44ec8ef35dd6ecbc9c6c9ab37b5e7f1ffa020af116d1d301f87791841cf1461c774e8ef44094e9ef4d3b1c5d5427628b3f1ff831466a880ff80ff8080",
      },
      "id": "0x44a1d78b820f6404de3cc45b34932178f9ac8f3d9114db279f657ca83fa751b7",
      "nft_detail": undefined,
      "nft_target": undefined,
      "nft_uri": "",
      "royalty": undefined,
      "target": "0xbae24162efbd568f89bc7a340798a6118df0189eb9e3f8697bcea27af99f8f79",
    },
  ],
  "requested": Array [
    Object {
      "amount": 100000000000n,
      "cat_target": undefined,
      "coin": Object {
        "coin": Object {
          "amount": 0n,
          "parent_coin_info": "0x0000000000000000000000000000000000000000000000000000000000000000",
          "puzzle_hash": "0xbae24162efbd568f89bc7a340798a6118df0189eb9e3f8697bcea27af99f8f79",
        },
        "puzzle_reveal": "0xff02ffff01ff02ff0affff04ff02ffff04ff03ff80808080ffff04ffff01ffff333effff02ffff03ff05ffff01ff04ffff04ff0cffff04ffff02ff1effff04ff02ffff04ff09ff80808080ff808080ffff02ff16ffff04ff02ffff04ff19ffff04ffff02ff0affff04ff02ffff04ff0dff80808080ff808080808080ff8080ff0180ffff02ffff03ff05ffff01ff04ffff04ff08ff0980ffff02ff16ffff04ff02ffff04ff0dffff04ff0bff808080808080ffff010b80ff0180ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff1effff04ff02ffff04ff09ff80808080ffff02ff1effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080",
        "solution": "0xffffa033e6b7efb2fa91742b9a4dc36fc96dc1255ced7fe78ffff469bd24aacefa913bffffa0f1c3875cc0f8cd4926d8a9423a244c30af5b7dfa7a25c947c570634423ca2157ff85174876e800ff80808080",
      },
      "id": "",
      "nft_detail": undefined,
      "nft_target": undefined,
      "nft_uri": "",
      "royalty": undefined,
      "target": "0xf1c3875cc0f8cd4926d8a9423a244c30af5b7dfa7a25c947c570634423ca2157",
    },
  ],
  "settlementModName": "settlement_payments",
}
`;
